### PR TITLE
GoodCigarFilter + tests

### PR DESCRIPTION
**Pull request by akiezun**
ported BadCigarFilter (in Hellbender filters have positive names, in concordance with Java filter semantics) + tests (added 2 tests to cover 2 more branches, 1 branch seems unreachable).

@vruano  please review
addresses https://github.com/broadinstitute/hellbender/issues/373

### CODE REVIEW

**Comment by akiezun** [comment]
@vruano is on vacation. @lbergelson please review

**Comment by vruano** [comment]
Done with my review. I'm to assigning it back to @akiezun in case @lbergelson wants to add.

**Comment by akiezun** [comment]
I think we should write most readable code for now and optimize with a profiler later. The regexp machinery in Java libraries definitely does a better job optimizing the state machine than a person can.

**Comment by vruano** [comment]
Finished with my review back to @akiezun

**Comment by vruano** [comment]
Please check on my general comments posted at the end of the second commit, I thought that they would be posted here as well.

**Comment by akiezun** [comment]
@vruano you're raising 3 issues:
1) is the algorithm correct?
2) is the algorithm's results the same as in GATK3?
3) is the algorithm esthetically pleasing/readable.

For 1, I'm going to create a new ticket - it's not under discussion here.
For 2, I'll enter the cases you're suggesting as unit tests and see what happens.
For 3, I can work on making it read better. 

My first goal in this tickets is to port the existing filter. If we can improve the readability at the same time, that's great. Fixing any bugs or deficiencies of the algorithm is strictly future work (if it has been broken all this time in GATK3, it's ok for it to be broken for a few more weeks in GATK4).

**Comment by akiezun** [comment]
@eitanbanks and @vdauwera I heard from @droazen that you guys are working on the BadCigarFilter in GATK3. Can you summarize the work? we need to synchronize the changes/improvements across the 2 codebases.

**Comment by eitanbanks** [comment]
BadCigarFilter was made a default Walker filter and functionality was added to it so that it now also filters out reads whose cigar length does not equal the read length.

**Comment by akiezun** [comment]
is the code in?

**Comment by vdauwera** [comment]
See https://github.com/broadinstitute/gsa-unstable/pull/930

The title is misleading -- see the last few comments in the conversation for a summary of the work done.

**Comment by akiezun** [comment]
why not rename the issue then?

**Comment by akiezun** [comment]
@eitanbanks thanks. the code and tests are now in hellbender too.

**Comment by akiezun** [comment]
@vruano can you give an example for "Also it does accept non empty CIGAR with out any operator that consume read bases such as N or P. That would not happen with the GATK3 code." ?

**Comment by akiezun** [comment]
@vruano back to you

I fixed the 1H1H1M problem you brought up "GATK3's filter would not complain for consecutive H operations at the beginning whereas the new code does" and I entered tests for it (one for head, one for tail).

I entered https://github.com/broadinstitute/hellbender/issues/428 for clarifying the spec.

**Comment by vruano** [comment]
What is that for example "10P" or "101N" would fail in the GATK3 code but the would not in the current pull-request. The reason is that the P and N operators return false for the method "consumesReadBases()" and the GATK3 code pays attention to that.

File: src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * From SAM spec:
+ *  - H can only be present as the first and/or last operation
+ *  - S may only have H operations between them and the ends of the CIGAR string
+ *
+ *  ps: apparently an empty cigar is valid.
+ */
+public final class GoodCigarFilter implements ReadFilter {
+
+    private static final String DIGITS = "\\d*"; //Any operator can be preceeded by a series of digits.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java]: This is incorrect -- each operator must be preceded by a series of digits. The regexp for an entire CIGAR string (from the SAM spec) is: `\*|([0-9]+[MIDNSHPX=])+`

File: src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    //We pre-create one regular expression pattern and reuse it.
+    //Each clause in the pattern corresponds to one condition we're testing for.
+    private final static Pattern BAR_CIGAR = Pattern.compile(
+           any((ANYTHING + NOT_H + H + NOT_H + ANYTHING),                      //H in the middle without flanking H
+               (ANYTHING + NOT_H + S + NOT_H + ANYTHING),                      //S in the middle without flanking H
+               (ANYTHING + any(I+I, D+D, I+D, D+I) + ANYTHING),                //consecutive indels
+               (maybeSome(CLIP)+ D + ANYTHING),                                //starts with deletion
+               (ANYTHING + D + maybeSome(CLIP)),                               //ends with deletion
+               (some(CLIP))));                                                 //only clips
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java]: This approach makes me incredibly nervous -- complex regular expressions are notoriously prone to edge conditions and hard-to-see bugs, especially when (as here) there are multiple indefinite quantifiers. My experience has been that regular expressions with more than 2-3 `*` and/or `+` quantifiers tend to be hopelessly brittle and prone to weird results. Eg., are you ABSOLUTELY sure that you want to be always using greedy quantifiers (`*` and `+`) instead of reluctant quanitifiers (`*?` and `+?`) in your `maybeSome()` and `some()` methods? Have you thought about how these interact in every scenario? I can almost guarantee that there is at least one case where a `*` or `+` will eat too much of the cigar and lead to some erroneous result...

I feel that a straightforward parsing of the Cigar elements would be far less error-prone.

File: src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    private static final String EQ= cat(DIGITS, CigarOperator.EQ.name());
+    private static final String X = cat(DIGITS, CigarOperator.X.name());
+
+    private static final String NOT_H    = any(M, S, I, D, N, P, EQ, X);
+    private static final String ANYTHING = maybeSome(any(NOT_H, H));
+    private static final String CLIP     = any(H, S);
+
+    //We pre-create one regular expression pattern and reuse it.
+    //Each clause in the pattern corresponds to one condition we're testing for.
+    private final static Pattern BAR_CIGAR = Pattern.compile(
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java]: Should be named `BAD_CIGAR`, not `BAR_CIGAR`

**Comment by vruano** [comment]
Also it seems that several soft clips are allowed in GATK3 code as long as is before or after the non-clipping core section of the CIGAR.

So the following would pass the filter:

1S20S10M 
1H1S10M10S30H
1H1S10M10S1S30H

But these wouldn't:

1S1H10M
1H1M1S1M
1M1H1S

**Comment by vruano** [comment]
Is much the same as with H.

**Comment by vruano** [comment]
Done with the review.

**Comment by akiezun** [comment]
@vruano to clarify. both 10P and 101N pass on GATK3. I just checked.

**Comment by akiezun** [comment]
by "pass" i mean - they are not flagged as Bad

File: src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/filters/GoodCigarFilter.java]: True. A cigar string without a number is malformed so you can't make a samrecord. In the regexp setting you can at least test the string itself and validate that if there ever were a read with this cigar we would reject it. Anyway, fixed.

**Comment by akiezun** [comment]
alright @vruano @droazen  
as discussed today I will recode this as a bunch of loops - the GATK3 code is atrocious but my regexp did not me win the fan following i was going for. And ultimately it's about the readability at this point (which is a subjective manner, you guys said loops would make it more readable and people's opinion defines 'readability')

**Comment by vruano** [comment]
"@vruano to clarify. both 10P and 101N pass on GATK3. I just checked."

I see... it seems that this might be cause by yet another bug in GATK3 code ... but I guess that is to be left for the other issue.

From this code within the second loop:

```java
if (!hasMeaningfulElements && op.consumesReadBases()) {
                    hasMeaningfulElements = true;
}
```
one would think that to have a meaningful-element is to have one that consumes read bases.

However it turns out that this if condition is never true as:

```java
boolean hasMeaningfulElements = (firstOp != CigarOperator.H && firstOp != CigarOperator.S);
```
is guaranteed to initialize it to true. 

So that is some code you can get rid off for now. But I guess that it need to be fixed eventually.

**Comment by akiezun** [comment]
Does anyone know why there are 3 implementation of validity for Cigars in GATK3?
1) in the BadCigarFilter
2) in CigarUtils
3) in Cigar itself ?

**Comment by akiezun** [comment]
the third one is in htsjdk obviously. anyway, i'm merging the first two

**Comment by akiezun** [comment]
I clarified the semantics of Good CIGAR. It does not make sense to have 'good cigars' that are invalid according to htsjdk so I clarified the semantics to state that a "good" cigar is a valid cigar that passes additional criteria:
```
  - does not have consecutive I/D elements
  - does not start or end with deletions (with or without preceding clips).
```
(now, we can argue about those criteria but those are the criteria in GATK3 and the current code and doc makes this explicit.)

I rewrote the whole code again to reflect this change. @vruano back to you.

**Comment by vruano** [comment]
The code is far easier to understand now. Mostly minor changes and a suggestion to improve the readability of one of the supporting private methods. 

CigarUtils.countRefBasesBasedOnCigar have some issues but these are not really part of your code so it might be fair to address them separately in another pull-request. 

Back to @akiezun.

**Comment by akiezun** [comment]
This filter did not deserve a special file after I moved the isGood method to CigarUtils. So I moved this filter to ReadFilterLibrary and moved tests accordingly.

back to @vruano

**Comment by vruano** [comment]
@akiezun please rebase, squash and merge.
